### ISSUE TITLE
Equality narrowing for strings/ints/bytes is inconsistent with reachability analysis

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#4051]
### Summary

The first function here causes us to emit an `invalid-return-type` diagnostic: our reachability analysis determines that there are possible paths through the function that lead to the function implicitly returning `None`. The second function does not, however: the `reveal_type` call at the end of the function is inferred as returning `Never` (because the type of `x` at that point in the code is inferred as being `Never`), adding an additional constraint to the control-flow graph and leading ty to conclude that there is no path through the function that would lead to ty implicitly returning `None`: 

```py
def f(x: str) -> int:  # error: [invalid-return-type]
    if x == 'foo':
        return 42
    elif x != 'foo':
        return 42


def g(x: str) -> int:  # no error
    if x == 'foo':
        return 42
    elif x != 'foo':
        return 42

    reveal_type(x)  # revealed: Never
```

This feels pretty confusing and inconsistent. In runs of ty where `strict-literal-narrowing` is set to `false`, I think we have to try to have our reachability analysis match our type-narrowing analysis. The implicit `else` branch after the `elif` branch (the implicit fallthrough branch taken if `x != 'foo' and x == 'foo'`) should be understood as unreachable when this mode (now the default mode of ty) is enabled. That would lead us to accept both functions when `strict-literal-narrowing` is set to `false` and reject both functions when `strict-literal-narrowing` is set to `true`.

**Comment by kimjune01** [comment]
Reversing the two conditions makes this go away, with no `reveal_type` involved:

```py
def eq_first(x: str) -> int:  # error: [invalid-return-type]
    if x == "foo":
        return 42
    elif x != "foo":
        return 42

def ne_first(x: str) -> int:  # no error
    if x != "foo":
        return 42
    elif x == "foo":
        return 42
```

Reachability already reads the narrowed type in both, so what differs is whether the residue makes the second comparison decidable: `Literal["foo"]` gives `Literal[True]`, while `str & ~Literal["foo"]` gives `bool`, which is the "Negative contributions" gap in `comparison/intersections.md`.
